internal/server/multitherading.go

```diff
@@ -0,0 +1,120 @@
+package server
+
+import (
+	"bytes"
+
+	"github.com/dicedb/dice/internal/cmd"
+	"github.com/dicedb/dice/internal/comm"
+	"github.com/dicedb/dice/internal/eval"
+	"github.com/dicedb/dice/internal/ops"
+	"github.com/dicedb/dice/internal/shard"
+	"github.com/twmb/murmur3"
+)
+
+func getShard(key string, n uint32) uint32 {
+	// Hash the key using MurmurHash3
+	hash := murmur3.Sum32([]byte(key))
+	// Apply modulo operation
+	return hash % n
+}
+
+type ShardingMeta struct {
+	Cmd     string
+	Breakup func(mgr *shard.ShardManager, redisCmd *cmd.RedisCmd, c *comm.Client) []cmd.RedisCmd
+	Gather  func(responses ...eval.EvalResponse) []byte
+}
+
+var (
+	MultithreadingCmds = map[string]ShardingMeta{}
+
+	pingCmdMeta = ShardingMeta{
+		Cmd:     "PING",
+		Breakup: scatterPING,
+		Gather:  gatherPING,
+	}
+)
+
+func init() {
+	MultithreadingCmds["PING"] = pingCmdMeta
+}
+
+func (s *AsyncServer) cmdsBreakup(redisCmd *cmd.RedisCmd, c *comm.Client) []cmd.RedisCmd {
+
+	// check if command supports multisharding
+	val, ok := MultithreadingCmds[redisCmd.Cmd]
+	// If not then just return the command as it is
+	// no need to do futher breakup
+	if !ok {
+		return []cmd.RedisCmd{*redisCmd}
+	}
+
+	// If yes then send the command to the respective breakup function
+	// Which can return array of broken down commands
+	return val.Breakup(s.shardManager, redisCmd, c)
+}
+
+func (s *AsyncServer) scatter(cmds []cmd.RedisCmd, c *comm.Client) {
+	// single sharded command
+	if len(cmds) == 1 {
+		key := cmds[0].Args[0]
+		id := getShard(key, uint32(s.shardManager.GetShardCount()))
+		s.shardManager.GetShard(shard.ShardID(id)).ReqChan <- &ops.StoreOp{
+			Cmd:      &cmds[0],
+			WorkerID: "server",
+			ShardID:  int(id),
+			Client:   c,
+		}
+	} else {
+		// multishard command
+		// Condition for command that requires all shards such as PING
+		if len(cmds) == s.shardManager.GetShardCount() {
+			for i := 0; i < len(cmds); i++ {
+				s.shardManager.GetShard(shard.ShardID(i)).ReqChan <- &ops.StoreOp{
+					Cmd:      &cmds[i],
+					WorkerID: "server",
+					ShardID:  i,
+					Client:   c,
+				}
+			}
+		} else {
+			// Otherwise check for the shard based on the key using hash
+			// and send it to the particular shard
+			for i := 0; i < len(cmds); i++ {
+				key := cmds[i].Args[0]
```

Comment:
This may not work for multi-key commands.
Before this we need to break down the redis command to store operations.
RESP parser would return an array of REDIS(client) commands.